### ISSUE TITLE
[improvement]: Security Hardening - Running Asterisk as non-root 

### ISSUE THREAD

**Issue by TheRook** [asterisk/asterisk#333]
### Improvement Description

I have found that Asterisk tends to run with root permissions on many deployments, and while this was once common place - it violates the principle of least privileged access.  If [deployed properly](http://www.asteriskdocs.org/en/2nd_Edition/asterisk-book-html-chunk/asterisk-CHP-13-SECT-4.html) asterisk doesn't appear to need root access at all because it doesn't bind to a port below tcp/1024.

Given the context of root permissions, if you look at the three most recent [Asterisk Security Advisories](https://www.asterisk.org/downloads/security-advisories/) (GHSA-4xjp-22g4-9fxm AST-2022-009 and AST-2022-008) we can see that memory corruption vulnerabilities in a root process increases its CVSS score as this would result in a remote root compromise.  AST-2022-009 Is a file disclosure vulnerability, and this is of particular interest - in this context giving root access to Asterisk means that it can be used to read /etc/shadow to gain credentials to login via SSH.  This turns a simple file disclosure vulnerability, or directory traversal or XXE - into a full remote compromise, which is a a very good reason not to hand out root access.

There are is also a [local privilege escalation vulnerability in Asterisk's](https://github.com/asterisk/asterisk/issues/316) via local file permissions.   In this attack pattern any directory that is in the $PATH, such as like $ASTVARRUNDIR could be used to introduce a new binary that is run as root permissions by asterisk or as by the native root user.

Looking back and seeing how root permissions creates local privilege escalation opportunities and remote code execution, i think it is in everyone's best interest that we migrate Asterisk to a principle of least privilege access.  I'd like to see an announcement where we notify the community that in an effort of security modernization Asterisk will refuse to run as root.   Then a CVE can be issued so that package maintainers know to patch a specific issue.

Finally, protective code should be added to asterisk to prevent it from ever running as root.
```
    if (geteuid() == 0) {
        printf("Dropping root privileges...\n");
        
        // Try to set the effective user ID to a non-root user
        if (seteuid(1000) != 0) { // Change "1000" to the desired non-root user's UID
            perror("seteuid");
            exit(EXIT_FAILURE);
        }
```
In summation, if hardening guidelines where followed then AST-2022-009 wouldn't lead to RCE, and issues like #316 wouldn't be exploitable.  This is why its important to follow security best practices, and in this case we can see that many package maintainers have not updated to follow modern security guidelines.  It is in the best interest of users and companies that use this software to make sure that it is deployed in a secure way, and refusing to run as root will make sure that Asterisk deployments are safe.

**Comment by TheRook** [comment]
Also, 17 cases of `0777` world read/write/execute isn't great. Files like logs should probably be 744 for read only, and 755 for directories that need read+execute:
https://github.com/search?q=repo%3Aasterisk%2Fasterisk+0777&type=code

**Comment by TheRook** [comment]
Do I need to submit a PR and contact Mitre for a CVE?  What is the next step here?

**Comment by jcolp** [comment]
The issue has been triaged, accepted, and is in queue. Unless you are familiar with Asterisk itself, I wouldn't recommend making changes because fundamentally altering behavior (such as not allowing root) even in this case is going to require discussion, research, and understanding all of the repercussions. If a CVE is warranted in some way then it will be requested when this work is undertaken.

**Comment by TheRook** [comment]
I exploited asterisk in a pentest to get root, dump creds and pivot.  This
bug was used to hack a bank.

A CVE should be issued.

**Comment by jcolp** [comment]
This is the first time you have mentioned at all that you have identified an actual exploitable method based on the security information and practices you've been talking about. Please use the Security Vulnerability report process of the project to privately report the details:

https://github.com/asterisk/asterisk/security/advisories/new

It will then be triaged, identified, scheduled, and a CVE issued when a release occurs.

A public Github issue is not the location to discuss any further detail.

**Comment by dovi5988** [comment]
@jcolp There is other software (such as VLC) which by default will not let you run it as root. A compromise could be to only allow Asterisk to run as root *if* the user passes a flag to allow to run as root even though it's highly not recommended.

**Comment by jcolp** [comment]
@dovi5988 Yes, there are options and possibilities.

**Comment by TheRook** [comment]
@jcolp Sorry for the delay on the writeup, I am installing the latest version and getting a proper writeup so that we can have a good discussion. 

@dovi5988  A security override this doesn't bother me, I have seen this in other products. As long as it defaults to a safe condition and writes in the log each time that it is deployed in an insecure way - then there isn't a chance of a silent bug. The point of the log write is for a large orgization that uses ELK or another log management system, they can track security anomalies, and this fits.

**Comment by sosnik** [comment]
Any update on this?

**Comment by jcolp** [comment]
No, there is no update. Any updates will be posted on here, and any PRs referencing this will be linked here.
